lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -891,6 +891,104 @@ OpFoldResult OrOp::fold(ArrayRef<Attribute> constants) {
       constants, [](APInt &a, const APInt &b) { a |= b; });
 }
 
+/// Simplify concat ops in an or op when a constant operand is present in either
+/// concat.
+///
+/// This will invert an or(concat, concat) into concat(or, or, ...), which can
+/// often be further simplified due to the smaller or ops being easier to fold.
+///
+/// For example:
+///
+/// or(..., concat(x, 0), concat(0, y))
+///    ==> or(..., concat(x, 0, y)), when x and y don't overlap.
+///
+/// or(..., concat(x: i2, cst1: i4), concat(cst2: i5, y: i1))
+///    ==> or(..., concat(or(x: i2,               extract(cst2, 4..3)),
+///                       or(extract(cst1, 3..1), extract(cst2, 2..0)),
+///                       or(extract(cst1, 0..0), y: i1))
+static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
+                                                   size_t concatIdx2,
+                                                   PatternRewriter &rewriter) {
+  assert(concatIdx1 < concatIdx2 && "concatIdx1 must be < concatIdx2");
+
+  auto inputs = op.inputs();
+  auto concat1 = inputs[concatIdx1].getDefiningOp<ConcatOp>();
+  auto concat2 = inputs[concatIdx2].getDefiningOp<ConcatOp>();
+
+  assert(concat1 && concat2 && "expected indexes to point to ConcatOps");
+
+  // We can simplify as long as a constant is present in either concat.
+  bool hasConstantOp1 =
+      llvm::any_of(concat1->getOperands(), [&](Value operand) -> bool {
+        return operand.getDefiningOp<hw::ConstantOp>();
+      });
+  bool hasConstantOp2 =
+      llvm::any_of(concat2->getOperands(), [&](Value operand) -> bool {
+        return operand.getDefiningOp<hw::ConstantOp>();
+      });
+  if (!hasConstantOp1 && !hasConstantOp2)
```

Comment:
Random drive-by, but I'd move the short-circuit up.  If hasConstantOp1 is true, then no need to compute hasConstantOp2.  Something like:

if (!hasConstantOp1) {
  if (!anyof(...))
    return false;
}

would look nice.

Reply:
Sounds good, I can quickly make this change, and will keep this kind of thing in mind in the future.

Reply:
I ended up committing this directly to main in 00a3c0d723605e4a932f55187007bd98cd94a30b, which I'm assuming was OK for this kind of change, based on what I'm reading in the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access), but let me know if I should submit pull requests for this kind of thing in the future.

Reply:
Thanks!  LGTM